Fix the TimePicker hour/minute field's input actions

TahaTesser opened the pull request:
fixes https://github.com/flutter/flutter/issues/68249

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: const TimePickerSample(),
    );
  }
}

class TimePickerSample extends StatelessWidget {
  const TimePickerSample({Key? key}) : super(key: key);

  Future<TimeOfDay?> _showTimePicker(BuildContext context) {
    return showTimePicker(
      initialTime: TimeOfDay.now(),
      initialEntryMode: TimePickerEntryMode.input,
      context: context,
      confirmText: 'Accept',
      cancelText: 'Reject',
      hourLabelText: 'Happy Hour',
      minuteLabelText: 'Happy Minute',
      helpText: 'Spend your time wisely',
      onEntryModeChanged: (TimePickerEntryMode mode) {
        // ignore: avoid_print
        print('Current Mode: $mode');
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('TimePicker Sample'),
      ),
      body: Center(
        child: OutlinedButton(
          onPressed: () => _showTimePicker(context),
          child: const Text('Show TimePicker'),
        ),
      ),
    );
  }
}
``` 
	
</details>

### Native behavior with input mode 
The hour text field has the "next" input action which shifts the focus from the hour field to the minute field.
The minute text field has the "done" input action, which hides the keyboard. 

https://user-images.githubusercontent.com/48603081/151339372-3e952ca4-df9c-459d-9ede-263fcd253349.mov

### Bug
Both hour and minute text fields have "done" action which when pressing does nothing.




https://user-images.githubusercontent.com/48603081/166164962-aac21ecc-81b1-4d34-8055-ed4e52650db7.mov


### Fix


https://user-images.githubusercontent.com/48603081/166164973-beeb2af6-9f1e-42b6-b6be-cbbcb93a255c.mov



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1339,6 +1339,21 @@ class _TimePickerInputState extends State<_TimePickerInput> with RestorationMixi
   late final RestorableTimeOfDay _selectedTime = RestorableTimeOfDay(widget.initialSelectedTime);
   final RestorableBool hourHasError = RestorableBool(false);
   final RestorableBool minuteHasError = RestorableBool(false);
+  late FocusNode _hourFocusNode;
```
Do we really need to manage these focus nodes ourselves? If we need to transfer focus to the next node, we could just use something like: `FocusScope.of(context).nextFocus()`.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1339,6 +1339,21 @@ class _TimePickerInputState extends State<_TimePickerInput> with RestorationMixi
+  @override
+  void initState() {
+    super.initState();
+    _hourFocusNode = FocusNode()..addListener(() {
+      setState(() {
+      });
+    });
+    _minuteFocusNode = FocusNode()..addListener(() {
+      setState(() {
+      });
+    });
```
Why are we calling `setState` when the focus node changes? The fields themselves should automatically rebuild as necessary.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1398,6 +1413,7 @@ class _TimePickerInputState extends State<_TimePickerInput> with RestorationMixi
       _selectedTime.value = TimeOfDay(hour: newHour, minute: _selectedTime.value.minute);
       widget.onChanged(_selectedTime.value);
     }
+    FocusScope.of(context).requestFocus(_minuteFocusNode);
```
Again, I think this could just be: 

```dart
FocusScope.of(context).nextFocus()
```

and should probably be moved into the if statement above as if we don't have a valid hour we should stay put.

TahaTesser replied:
`FocusScope.of(context).nextFocus().` simply unfocus the hour field hence I am requesting minute focus.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1414,6 +1430,7 @@ class _TimePickerInputState extends State<_TimePickerInput> with RestorationMixi
       _selectedTime.value = TimeOfDay(hour: _selectedTime.value.hour, minute: int.parse(value!));
       widget.onChanged(_selectedTime.value);
     }
+    _minuteFocusNode.unfocus();
```
Is this used to dismiss the keyboard? If so, then we could probably just:

```dart
FocusScope.of(context).unfocus();
```

and again move it into the if statement.

darrenaustin reviewed (changes requested):
@TahaTesser thanks for the contribution. 

This looks reasonable, but I think it could be simplified by not creating and managing the focus nodes. in the `_TimePickerInputState`. See my comments below. Thx.

TahaTesser committed:
Fix the TimePicker hour/minute field's input actions

TahaTesser commented:
@darrenaustin 
Removed the focus nodes added earlier, moved the Focus score code into conditions, and updated the test and PR description. 

Can you please take a look?  Thank you!

guidezpl reviewed (approved):
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

Any objections @darrenaustin ?

TahaTesser commented:
Waiting for @darrenaustin's confirmation too before merging this, thanks.

darrenaustin reviewed (approved):
Yikes, so sorry for the lag on this. LGTM. Thx!
